enhancements/network/bgp-ovn-kubernetes.md

```diff
@@ -0,0 +1,701 @@
+
+As an alternative, configuring BGP peers and which route-targets to import would eliminate the need to manually
+configure routes in the host, and would allow dynamic routing updates based on changes in the provider’s network.
+
+### Exporting Routes into the Provider Network
+
+There exists a need for provider networks to learn routes directly to services and pods today in Kubernetes. More
+specifically, MetalLB is already one solution where load balancer IPs are advertised by BGP to provider networks. The
+goal of this RFE is to not duplicate or replace the function of MetalLB. MetalLB should be able to interoperate with
+OVN-Kubernetes, and be responsible for advertising services to a provider’s network.
+
+However, there is an alternative need to advertise pod IPs on the provider network. One use case is integration with 3rd
+party load balancers, where they terminate a load balancer and then send packets directly to OCP nodes with the
+destination IP address being the pod IP itself. Today these load balancers rely on custom operators to detect which node
+a pod is scheduled to and then add routes into its load balancer to send the packet to the right node.
+
+By integrating BGP and advertising the pod subnets/addresses directly on the provider network, load balancers and other
+entities on the network would be able to reach the pod IPs directly.
+
+### Datapath Performance
+
+In cases where throughput is a priority, using the underlay directly can eliminate the need for tunnel encapsulation,
+and thus reducing the overhead and byte size of each packet. This allows for greater throughput.
+
+### Multi-homing, Link Redundancy, Fast Convergence
+
+BGP can use multi-homing with ECMP routing in order to provide layer 3 failover. When a link goes down, BGP can reroute
+via a different path. This functionality can be coupled with BFD in order to provide fast failover.
+
+### User Stories
+
+ * As a user I want to be able to leverage my existing BGP network to dynamically learn routes to pods in my Kubernetes
+   cluster.
+ * As a user, rather than having to maintain routes with NM State manually in each Kubernetes node, as well as being
+   constrained to using local gateway mode for respecting user defined routes; I want to use BGP so that I can dynamically
+   advertise egress routes for the Kubernetes pod traffic in either gateway mode.
+ * As a user where maximum throughput is a priority, I want to reduce packet overhead by not having to encapsulate
+   traffic with Geneve.
+ * As a baremetal or egress IP user, I do not want to have to restrict my nodes to the same layer 2 segment and prefer
+   to use a pure routing implementation to handle advertising virtual IP (VIP) movement across nodes.
+
+### Goals
+
+* To provide a user facing API to allow configuration of iBGP or eBGP peers, along typical BGP configuration including
+  communities, route filtering, etc.
+* Support for advertising Egress IP addresses.
+* To enable BFD to BGP peers.
```

Comment:
I don't know a lot about BGP and I know even less about BFD, but this feels underspecified as well. Among other things, does this mean allowing ovn-k to do BFD to external BGP peers, or allowing the external network to do BFD to ovn-k? And why? What will this enable that wouldn't be possible without this feature?

Reply:
A BGP session can be associated and backed up by a BFD session towards the same neighbor.
BGP alone has a min link failure detection time of 3 seconds, while bfd can go sub second.
When this is enabled, as soon as bfd declares the link dead, the bgp session disconnects and the routes through that link are removed.

Reply:
The question of "what" is doing BFD is s good one. OVN is capable of doing BFD. We use it today with Multiple External Gateways (MEG). In this case when one of the external gateways goes down, BFD detects it within a couple hundred milliseconds, and OVN purges the routes to that gateway, so that egress traffic will choose  a different one.

With FRR, it has BFD daemon support, where it can run BFD in the host and detect if peers go down. With this enhancement OVN has no integration with FRR, so it cannot tell FRR to remove the routes from its RIB (although it could remove the routes from the OVN datapath). Trying to hack that together would be dicey until we have real support with OVN and FRR. Therefore for the purpose of this enhancement, the BFD will be done with FRR so the flow would look like this:

1. BFDd detects a failure and notifies FRR.
2. FRR removes the routes from its RIB.
3. OVNK notices via netlink that the routes are gone, and removes them from OVN.

Reply:
added a BFD section in the doc